core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/JniPCConnector.java

```diff
@@ -126,15 +140,94 @@ private static boolean initNativeCode() {
         return true;
     }
 
-    private static boolean loadLibrary() {
+    /**
+     * The method will try to extract the dll for the Windows performance counters to a local
+     * folder and then will try to load it. The method will do all that by doing the following things:
+     * 1. Find the OS type (64/32) currently supports only 64 bit.
+     * 2. Will find the path to extract to, which is %temp%/sdk_version_number
+     * 3. Find out whether or not the file is already exists in that directory
+     * 4. If the dll is not there, the method will extract it from the jar to that directory
+     * 5. The method will call System.load to load the dll and by doing so we are ready to use it
+     * @return true on success, otherwise false
+     * @throws IOException If there are errors in opening/writing/reading/closing etc.
+     *         Note that the method might throw RuntimeExceptions due to critical issues
+     */
+    private static boolean loadLibrary() throws IOException {
         String model = System.getProperty("sun.arch.data.model");
         String libraryToLoad = NATIVE_LIBRARY_64;
         if (!BITS_MODEL_64.equals(model)) {
             return false;
         }
 
-        System.loadLibrary(libraryToLoad);
+        File dllPath = buildDllLocalPath();
+
+        File dllOnDisk = new File(dllPath + File.separator + libraryToLoad);
+
+        if (!dllOnDisk.exists()) {
+            extractToLocalFolder(dllOnDisk, libraryToLoad);
+        }
+
+        System.load(dllOnDisk.toString());
+
         InternalLogger.INSTANCE.trace("Successfully loaded library '%s'", libraryToLoad);
+
         return true;
     }
+
+    private static void extractToLocalFolder(File dllOnDisk, String libraryToLoad) throws IOException {
```

Comment:
I think that it might be useful having a helper class/method for extracting a resource into a given folder instead of having it embedded in this class. WDYT?